File: dwarf/targets/cardano-node/src/DecodePeerSharing.hs
{-# LANGUAGE OverloadedStrings #-}
-- | cardano-node-mini-protocol-decode-peersharing shim.
-- Decodes one PeerSharing mini-protocol message envelope.
module Main (main) where

import qualified Codec.CBOR.Read                  as CBOR
import qualified Codec.CBOR.Term                  as CBOR
import qualified Data.ByteString.Lazy             as BSL
import           Data.Word                        (Word32)
import           System.Exit                      (ExitCode (..), exitSuccess, exitWith)

termList :: CBOR.Term -> Either String [CBOR.Term]
termList (CBOR.TList xs) = Right xs
termList (CBOR.TListI xs) = Right xs
termList other = Left ("expected list term, got " ++ show other)

termWord8 :: String -> CBOR.Term -> Either String Word
termWord8 name (CBOR.TInt n)
  | n >= 0 && n <= 255 = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word8 range: " ++ show n)
termWord8 name (CBOR.TInteger n)
  | n >= 0 && n <= 255 = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word8 range: " ++ show n)
termWord8 name other = Left (name ++ " expected unsigned integer, got " ++ show other)

termWord16 :: String -> CBOR.Term -> Either String Word
termWord16 name (CBOR.TInt n)
  | n >= 0 && n <= 65535 = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word16 range: " ++ show n)
termWord16 name (CBOR.TInteger n)
  | n >= 0 && n <= 65535 = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word16 range: " ++ show n)
termWord16 name other = Left (name ++ " expected unsigned integer, got " ++ show other)

termWord32 :: String -> CBOR.Term -> Either String Word32
termWord32 name (CBOR.TInt n)
  | n >= 0 && n <= fromIntegral (maxBound :: Word32) = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word32 range: " ++ show n)
termWord32 name (CBOR.TInteger n)
  | n >= 0 && n <= fromIntegral (maxBound :: Word32) = Right (fromIntegral n)
  | otherwise = Left (name ++ " out of Word32 range: " ++ show n)
termWord32 name other = Left (name ++ " expected unsigned integer, got " ++ show other)

validateAddress :: CBOR.Term -> Either String ()
validateAddress term = do
  xs <- termList term
  case xs of
    [keyTerm, ipv4, port] -> do
      key <- termWord16 "peer address key" keyTerm
      if key == 0
         then termWord32 "ipv4 word" ipv4 >> termWord16 "ipv4 port" port >> Right ()
         else Left ("unexpected peer address key " ++ show key)
    [keyTerm, word1, word2, word3, word4, port] -> do
      key <- termWord16 "peer address key" keyTerm
      if key == 1
         then do
           _ <- termWord32 "ipv6 word1" word1
           _ <- termWord32 "ipv6 word2" word2
           _ <- termWord32 "ipv6 word3" word3
           _ <- termWord32 "ipv6 word4" word4
           _ <- termWord16 "ipv6 port" port
           Right ()
         else Left ("unexpected peer address key " ++ show key)
    _ -> Left ("unexpected peer address length " ++ show (length xs))

validate :: CBOR.Term -> Either String ()
validate term = do
  xs <- termList term
  case xs of
    [keyTerm, amount] -> do
      key <- termWord16 "message key" keyTerm
      case key of
        0 -> termWord8 "share amount" amount >> Right ()
        1 -> mapM_ validateAddress =<< termList amount
        _ -> Left ("unexpected peersharing message key " ++ show key)
    [keyTerm] -> do
      key <- termWord16 "message key" keyTerm
      if key == 2
         then Right ()
         else Left ("unexpected peersharing message key " ++ show key)
    _ -> Left ("unexpected peersharing message length " ++ show (length xs))

main :: IO ()
main = do
  bytes <- BSL.getContents
  case CBOR.deserialiseFromBytes CBOR.decodeTerm bytes of
    Left err -> do
      putStrLn ("ERR " ++ show err)
      exitWith (ExitFailure 1)
    Right (rest, term) | BSL.null rest ->
      case validate term of
        Right () -> do
          putStrLn "OK"
          exitSuccess
        Left msg -> do
          putStrLn ("ERR " ++ msg)
          exitWith (ExitFailure 1)
    Right _ -> do
      putStrLn "ERR trailing bytes after message"
      exitWith (ExitFailure 1)

File: dwarf/targets/amaru/src/bin/decode_peersharing.rs
// amaru-mini-protocol-decode-peersharing shim.
// Decodes one PeerSharing mini-protocol message envelope:
//   ShareRequest = [0, amount_u8]
//   SharePeers   = [1, [peer_address, ...]]
//   Done         = [2]
//
// The checked-out Amaru tree does not currently expose a PeerSharing codec
// under crates/. This shim follows Cardano PeerSharing wire evidence from
// ouroboros-network and the local Pallas PeerSharing codec, constrained to
// the cardano-node remote-address port width for cross-implementation checks.

use std::io::{self, Read};
use std::process::ExitCode;

use minicbor::data::Type;

fn read_list(
    d: &mut minicbor::Decoder<'_>,
    what: &str,
    mut item: impl FnMut(&mut minicbor::Decoder<'_>) -> Result<(), String>,
) -> Result<(), String> {
    match d.array().map_err(|e| format!("{what} array: {e}"))? {
        Some(len) => {
            for _ in 0..len {
                item(d)?;
            }
        }
        None => loop {
            if d.datatype().map_err(|e| format!("{what} item type: {e}"))? == Type::Break {
                d.skip().map_err(|e| format!("{what} break: {e}"))?;
                break;
            }
            item(d)?;
        },
    }
    Ok(())
}

fn decode_peer_address(d: &mut minicbor::Decoder<'_>) -> Result<(), String> {
    let len = d
        .array()
        .map_err(|e| format!("peer address array: {e}"))?
        .ok_or_else(|| "indefinite peer address not allowed".to_string())?;
    let key = d.u16().map_err(|e| format!("peer address key: {e}"))?;
    match (len, key) {
        (3, 0) => {
            let _ipv4 = d.u32().map_err(|e| format!("ipv4 word: {e}"))?;
            let _port = d.u16().map_err(|e| format!("ipv4 port: {e}"))?;
        }
        (6, 1) => {
            let _word1 = d.u32().map_err(|e| format!("ipv6 word1: {e}"))?;
            let _word2 = d.u32().map_err(|e| format!("ipv6 word2: {e}"))?;
            let _word3 = d.u32().map_err(|e| format!("ipv6 word3: {e}"))?;
            let _word4 = d.u32().map_err(|e| format!("ipv6 word4: {e}"))?;
            let _port = d.u16().map_err(|e| format!("ipv6 port: {e}"))?;
        }
        _ => {
            return Err(format!(
                "unexpected peer address (len, key) = ({len}, {key})"
            ))
        }
    }
    Ok(())
}

fn decode(data: &[u8]) -> Result<(), String> {
    let mut d = minicbor::Decoder::new(data);
    let len = d
        .array()
        .map_err(|e| format!("message array: {e}"))?
        .ok_or_else(|| "indefinite-length peersharing message not allowed".to_string())?;
    let key = d.u16().map_err(|e| format!("message key: {e}"))?;

    match (len, key) {
        (2, 0) => {
            let _amount = d.u8().map_err(|e| format!("share amount: {e}"))?;
        }
        (2, 1) => {
            read_list(&mut d, "share peers", decode_peer_address)?;
        }
        (1, 2) => {}
        _ => {
            return Err(format!(
                "unexpected peersharing message (len, key) = ({len}, {key})"
            ))
        }
    }

    if d.position() != data.len() {
        return Err(format!("trailing bytes at position {}", d.position()));
    }
    Ok(())
}

fn main() -> ExitCode {
    let mut buf = Vec::new();
    if let Err(e) = io::stdin().read_to_end(&mut buf) {
        println!("ERR stdin read: {e}");
        return ExitCode::from(1);
    }
    match decode(&buf) {
        Ok(()) => {
            println!("OK");
            ExitCode::from(0)
        }
        Err(msg) => {
            println!("ERR {msg}");
            ExitCode::from(1)
        }
    }
}
